deployment/common/changeset/link_transfer_timelock.go

```diff
@@ -0,0 +1,177 @@
+				return fmt.Errorf("value for transfers must be set")
+			}
+			if transfer.Value == big.NewInt(0) {
+				return fmt.Errorf("value for transfers must be non-zero")
+			}
+		}
+	}
+	// Check that Transfers and StartingOpCount have the same keys
+	for key := range cfg.Transfers {
+		_, err := chain_selectors.GetSelectorFamily(key)
+		if err != nil {
+			return fmt.Errorf("invalid chain selector: %w", err)
+		}
+	}
+	if !cfg.UseMCMS {
+		return nil
+	}
+	// Mcms specific configs
+	if cfg.McmsConfig == nil {
+		return fmt.Errorf("mcmsConfig must be set when UseMCMS is true")
+	}
+	// Upper bound for min delay (7 days)
+	if cfg.McmsConfig.MinDelay > 24*7*time.Hour {
+		return fmt.Errorf("minDelay must be less than 7 days")
+	}
+	return nil
+}
+
+// LinkTransferTimelock takes the given link transfers and executes them or creates an MCMS proposal for them.
+func LinkTransferTimelock(e deployment.Environment, req *LinkTransferTimelockConfig) (deployment.ChangesetOutput, error) {
+	err := req.Validate()
+	ctx := e.GetContext()
+	if err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("invalid LinkTransferTimelockConfig: %w", err)
+	}
+	chainSelectors := []uint64{}
+	for chainSelector := range req.Transfers {
+		chainSelectors = append(chainSelectors, chainSelector)
+	}
+	mcmsPerChain := map[uint64]*owner_helpers.ManyChainMultiSig{}
+
+	timelockAddresses := map[mcms.ChainIdentifier]common.Address{}
+	allBatches := []timelock.BatchChainOperation{}
+	for chainSelector := range req.Transfers {
+		chainID := mcms.ChainIdentifier(chainSelector)
+		chain := e.Chains[chainSelector]
+		addrs, err := e.ExistingAddresses.AddressesForChain(chainSelector)
+		linkState, err := MaybeLoadLinkTokenState(chain, addrs)
+		if err != nil {
+			return deployment.ChangesetOutput{}, err
+		}
+		linkAddress := linkState.LinkToken.Address()
+		mcmsState, err := MaybeLoadMCMSWithTimelockState(chain, addrs)
+		if err != nil {
+			return deployment.ChangesetOutput{}, err
+		}
+		timelockAddress := mcmsState.Timelock.Address()
+
+		mcmsPerChain[uint64(chainID)] = mcmsState.ProposerMcm
+
+		timelockAddresses[chainID] = timelockAddress
+		batch := timelock.BatchChainOperation{
+			ChainIdentifier: chainID,
+			Batch:           []mcms.Operation{},
+		}
+		opts := deployment.SimTransactOpts()
+		if !req.UseMCMS {
+			opts = chain.DeployerKey
+		}
+		totalAmount := big.NewInt(0)
+		for _, transfer := range req.Transfers[chainSelector] {
+			tx, err := linkState.LinkToken.Transfer(opts, transfer.To, transfer.Value)
+			if err != nil {
+				return deployment.ChangesetOutput{}, fmt.Errorf("error packing transfer tx data: %w", err)
+			}
+			op := mcms.Operation{
+				To:           linkAddress,
+				Data:         tx.Data(),
+				Value:        big.NewInt(0),
+				ContractType: string(types.LinkToken),
+			}
+			batch.Batch = append(batch.Batch, op)
+			totalAmount.Add(totalAmount, transfer.Value)
+		}
+		// check that from address has enough funds for the transfers
```

Comment:
this should be part of Validate before any onchain action happens otherwise you can have partial failures (say if MCMS config is false, then Transfer will take action onchain). I'd pass the env into your Validate function and ensure the current balance of the sender is greater than the total amount to transfer on each chain.

Reply:
Moved to the validate func, also added some new configs to be able to provide the `From` address so transfers are not restricted to timelock only https://github.com/smartcontractkit/chainlink/pull/15512/commits/6641f4f9112ee08257fa68e04bcc8d751cc649d8